### PR TITLE
Gh 9541 create mergeDiscriminatorSchema function

### PR DESCRIPTION
**Pull request by IslandRhythms**

### CODE REVIEW

File: lib/helpers/discriminator/mergeDiscriminatorSchema.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      continue;
+    }
+    if (omitNested[path]) {
+      continue;
+    }
+    if (specialProperties.has(key)) {
+      continue;
+    }
+    if (to[key] == null) {
+      to[key] = from[key];
```
[the remark this answers is not in the document]
Comment (hasezoey) [reply lib/helpers/discriminator/mergeDiscriminatorSchema.js:45]: i think this alert can be dismissed thanks to the check of [`specialProperties.has(key)`](https://github.com/Automattic/mongoose/pull/12452/files#diff-c1a625c1ae3ad53e2e66df88c3bb0ff96db3dd8e4cdca58d2ec8ca40275ada26R41-R43)

File: lib/helpers/discriminator/mergeDiscriminatorSchema.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+          }
+          continue;
+        } else if (isBsonType(from[key], 'ObjectID')) {
+          to[key] = new ObjectId(from[key]);
+          continue;
+        }
+      }
+      mergeDiscriminatorSchema(to[key], from[key], options, path ? path + '.' + key : key);
+    } else if (options.overwrite) {
+      to[key] = from[key];
```
[the remark this answers is not in the document]
Comment (hasezoey) [reply lib/helpers/discriminator/mergeDiscriminatorSchema.js:71]: i think this alert can be dismissed thanks to the check of [`specialProperties.has(key)`](https://github.com/Automattic/mongoose/pull/12452/files#diff-c1a625c1ae3ad53e2e66df88c3bb0ff96db3dd8e4cdca58d2ec8ca40275ada26R41-R43)

File: lib/helpers/discriminator/mergeDiscriminatorSchema.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Uzlopak) [reply lib/helpers/discriminator/mergeDiscriminatorSchema.js:45]: I agree

[the remark this answers is not in the document]
Comment (Uzlopak) [reply lib/helpers/discriminator/mergeDiscriminatorSchema.js:45]: Dismissed

File: lib/helpers/discriminator/mergeDiscriminatorSchema.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Uzlopak) [reply lib/helpers/discriminator/mergeDiscriminatorSchema.js:71]: Dismissed

**Comment by hasezoey** [comment]
should the option `isDiscriminatorSchemaMerge` now be removed from `utils.merge` in this PR too, or is it still used in other places?

File: lib/helpers/discriminator/mergeDiscriminatorSchema.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+  path = path || '';
+  const omitNested = options.omitNested || {};
+
+  while (i < len) {
+    key = keys[i++];
+    if (options.omit && options.omit[key]) {
+      continue;
+    }
+    if (omitNested[path]) {
```
Comment (vkarpov15) [inline lib/helpers/discriminator/mergeDiscriminatorSchema.js:38]: Get rid of `omitNested` and replace this check with `if (path === 'tree' && from != null && from.instanceOfSchema)`

File: lib/helpers/discriminator/mergeDiscriminatorSchema.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  if (from[trustedSymbol]) {
+    to[trustedSymbol] = from[trustedSymbol];
+  }
+
+  path = path || '';
+  const omitNested = options.omitNested || {};
+
+  while (i < len) {
+    key = keys[i++];
+    if (options.omit && options.omit[key]) {
```
Comment (vkarpov15) [inline lib/helpers/discriminator/mergeDiscriminatorSchema.js:35]: Remove the `omit` option and just check if `key` is either 'discriminators', 'base', or '_applyDiscriminators' here.

**Review by vkarpov15 — CHANGES_REQUESTED** [review]
Couple of suggestions to refactor. Ideally I'd like to call `mergeDiscriminatorSchema()` without any options.

**Review by vkarpov15 — APPROVED** [review]

**Pull request merged**